### PR TITLE
Document all the fields in the Cilium spec

### PR DESCRIPTION
**Pull request by olemarkus**
Since there is a massive amount of fields to document, I only added the docs to one of the packages so far. I will add to the other two if there is no problem with the docs I added here.

### CODE REVIEW

**Comment by johngmyers** [comment]
https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences

File: pkg/apis/kops/networking.go [hunk 1]
```diff
@@ -195,80 +195,224 @@ const CiliumIpamEni = "eni"
...
+	DisableK8sServices bool `json:"disableK8sServices,omitempty"`
+	// EnablePolicy determines the policy enforcement mode.
+	// "default": Follows Kubernetes policy enforcement.
+	// "always": Cilium restricts all traffic if no policy is in place.
+	// "never": Cilium allows all traffic regardless of policies in place.
+	EnablePolicy string `json:"enablePolicy,omitempty"`
+	// EnableTracing has not been implemented in the new templates and may be removed in the future.
+	// Setting this has currently no effect.
+	EnableTracing bool `json:"enableTracing,omitempty"`
+	// EnablePrometheusMetrics enables the Cilium /metrics endpoint for both the agent and the operator.
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]:
```suggestion
	// EnablePrometheusMetrics enables the Cilium "/metrics" endpoint for both the agent and the operator.
```

File: pkg/apis/kops/networking.go [hunk 2]
```diff
@@ -195,80 +195,224 @@ const CiliumIpamEni = "eni"
...
+	// Logstash has not been implemented in the new templates and may be removed in the future.
+	// Setting this has currently no effect.
+	Logstash bool `json:"logstash,omitempty"`
+	// LogstashAgent has not been implemented in the new templates and may be removed in the future.
+	// Setting this has currently no effect.
+	LogstashAgent string `json:"logstashAgent,omitempty"`
+	// LogstashProbeTimer has not been implemented in the new templates and may be removed in the future.
+	// Setting this has currently no effect.
+	LogstashProbeTimer uint32 `json:"logstashProbeTimer,omitempty"`
+	// DisableMasquerade determines if external traffic should be masqued behind the node IP.
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]:
```suggestion
	// DisableMasquerade disables masquerading traffic to external destinations behind the node IP.
```

File: pkg/apis/kops/networking.go [hunk 3]
```diff
@@ -195,80 +195,224 @@ const CiliumIpamEni = "eni"
...
 
-	AccessLog                string            `json:"accessLog,omitempty"`
-	AgentLabels              []string          `json:"agentLabels,omitempty"`
-	AgentPrometheusPort      int               `json:"agentPrometheusPort,omitempty"`
-	AllowLocalhost           string            `json:"allowLocalhost,omitempty"`
-	AutoIpv6NodeRoutes       bool              `json:"autoIpv6NodeRoutes,omitempty"`
-	BPFRoot                  string            `json:"bpfRoot,omitempty"`
-	ContainerRuntime         []string          `json:"containerRuntime,omitempty"`
+	// AccessLog has not been implemented in the new templates and may be removed in the future.
+	// Setting this has currently no effect.
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]:
```suggestion
	// AccessLog is not implemented and may be removed in the future.
	// Setting this has no effect.
```

(and in occurrences below)

File: pkg/apis/kops/networking.go [hunk 4]
```diff
@@ -195,80 +195,224 @@ const CiliumIpamEni = "eni"
...
+	// SocketPath has not been implemented in the new templates and may be removed in the future.
+	// Setting this has currently no effect.
+	SocketPath string `json:"socketPath,omitempty"`
+	// StateDir has not been implemented in the new templates and may be removed in the future.
+	// Setting this has currently no effect.
+	StateDir string `json:"stateDir,omitempty"`
+	// TracePayloadLen has not been implemented in the new templates and may be removed in the future.
+	// Setting this has currently no effect.
+	TracePayloadLen int `json:"tracePayloadlen,omitempty"`
+	// Tunnel determines the cilium tunelling mode. Possible values are "vxlan", "geneve", or "disabled".
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]:
```suggestion
	// Tunnel specifies the Cilium tunelling mode. Possible values are "vxlan", "geneve", or "disabled".
```
?

File: pkg/apis/kops/networking.go [hunk 5]
```diff
@@ -195,80 +195,224 @@ const CiliumIpamEni = "eni"
...
+	// Default: vxlan
+	Tunnel string `json:"tunnel,omitempty"`
+	// EnableIpv6 enables cluster ipv6 traffic. Note that if both EnableIpv6 and EnableIpv4 are set to false.
+	// Ipv4 will be enabled.
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]:
```suggestion
	// EnableIpv6 enables cluster IPv6 traffic. If both EnableIpv6 and EnableIpv4 are set to false
	// then IPv4 will be enabled.
```

File: pkg/apis/kops/networking.go [hunk 6]
```diff
@@ -195,80 +195,224 @@ const CiliumIpamEni = "eni"
...
+	// Default: false
+	EnableIpv6 bool `json:"enableipv6"`
+	// EnableIpv4 enables cluster ipv4 traffic. Note that if both EnableIpv6 and EnableIpv4 are set to false.
+	// Ipv4 will be enabled.
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]:
```suggestion
	// EnableIpv4 enables cluster IPv4 traffic. If both EnableIpv6 and EnableIpv4 are set to false
	// then IPv4 will be enabled.
```

File: pkg/apis/kops/networking.go [hunk 7]
```diff
@@ -195,80 +195,224 @@ const CiliumIpamEni = "eni"
...
+	// Default: false
+	EnableIpv4 bool `json:"enableipv4"`
+	// MonitorAggregation sets the level of packet monitoring. Can be set to either "low", "medium", or "maximum".
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]:
```suggestion
	// MonitorAggregation sets the level of packet monitoring. Possible values are "low", "medium", or "maximum".
```

File: pkg/apis/kops/networking.go [hunk 8]
```diff
@@ -195,80 +195,224 @@ const CiliumIpamEni = "eni"
...
+	// Default: medium
+	MonitorAggregation string `json:"monitorAggregation"`
+	// BPFCTGlobalTCPMax is the maximum number of entries in TCP CT table.
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]:
```suggestion
	// BPFCTGlobalTCPMax is the maximum number of entries in the TCP CT table.
```

File: pkg/apis/kops/networking.go [hunk 9]
```diff
@@ -195,80 +195,224 @@ const CiliumIpamEni = "eni"
...
+	// Default: 524288
+	BPFCTGlobalTCPMax int `json:"bpfCTGlobalTCPMax"`
+	// BPFCTGlobalAnyMax is the maximum number of entries in non-TCP CT table.
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]:
```suggestion
	// BPFCTGlobalAnyMax is the maximum number of entries in the non-TCP CT table.
```

File: pkg/apis/kops/networking.go [hunk 10]
```diff
@@ -195,80 +195,224 @@ const CiliumIpamEni = "eni"
...
+	// Default: 262144
+	BPFCTGlobalAnyMax int `json:"bpfCTGlobalAnyMax"`
+	// PreallocateBPFMaps reduce the per-packet latency at the expense of up-front memory allocation.
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]:
```suggestion
	// PreallocateBPFMaps reduces the per-packet latency at the expense of up-front memory allocation.
```

File: pkg/apis/kops/networking.go [hunk 11]
```diff
@@ -195,80 +195,224 @@ const CiliumIpamEni = "eni"
...
+	// Default: true
+	PreallocateBPFMaps bool `json:"preallocateBPFMaps"`
+	// SidecarIstioProxyImage is the regular expression matching compatible Istio sidecar istio-proxy.
+	// container image names.
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]:
```suggestion
	// SidecarIstioProxyImage is the regular expression matching compatible Istio sidecar istio-proxy
	// container image names.
```

File: pkg/apis/kops/networking.go [hunk 12]
```diff
@@ -195,80 +195,224 @@ const CiliumIpamEni = "eni"
...
+	// Default: cilium/istio_proxy
 	SidecarIstioProxyImage string `json:"sidecarIstioProxyImage"`
-	ClusterName            string `json:"clusterName"`
-	ToFqdnsEnablePoller    bool   `json:"toFqdnsEnablePoller"`
+	// ClusterName is the of the cluster and is only relevant when building a mesh of clusters.
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]:
```suggestion
	// ClusterName is the name of the cluster. It is only relevant when building a mesh of clusters.
```

File: pkg/apis/kops/networking.go [hunk 13]
```diff
@@ -195,80 +195,224 @@ const CiliumIpamEni = "eni"
...
+	ClusterName string `json:"clusterName"`
+	// ToFqdnsEnablePoller is disabled by default starting from version 1.4.x in favor
+	// of a more powerful DNS proxy-based implementation.
+	// Enable this option if you want to use FQDN policies but do not want to use
+	// the DNS proxy.
+	// To ease upgrade, users may opt to set this option to "true".
+	// Default: true
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]: Version 1.4.x of what? Kops? Cilium?

The default is "true"? That seems to contradict the previous text.

I suggest first saying what it does, then talk about what the default is and possibly why one would want to change it.

File: pkg/apis/kops/networking.go [hunk 14]
```diff
@@ -195,80 +195,224 @@ const CiliumIpamEni = "eni"
...
+	ToFqdnsEnablePoller bool `json:"toFqdnsEnablePoller"`
+	// ContainerRuntimeLabels determines the container runtime(s) used by Cilium
+	// Supported values are: "none", "containerd", "crio", "docker", "auto"
+	// Default: none
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]: If multiple values are possible, how are they given? Comma separated?

File: pkg/apis/kops/networking.go [hunk 15]
```diff
@@ -195,80 +195,224 @@ const CiliumIpamEni = "eni"
...
 	ContainerRuntimeLabels string `json:"containerRuntimeLabels,omitempty"`
-	IPTablesRulesNoinstall bool   `json:"IPTablesRulesNoinstall"`
-	AutoDirectNodeRoutes   bool   `json:"autoDirectNodeRoutes"`
-	EnableNodePort         bool   `json:"enableNodePort"`
-	Ipam                   string `json:"ipam,omitempty"`
+	// Ipam determines which IP address allocation mode to use.
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]:
```suggestion
	// Ipam determines the IP address allocation mode to use.
```
What are the other possible values? What is the default?

File: pkg/apis/kops/networking.go [hunk 16]
```diff
@@ -195,80 +195,224 @@ const CiliumIpamEni = "eni"
...
+	// "eni" will use AWS native networking for pods
+	Ipam string `json:"ipam,omitempty"`
+	// IPTablesRulesNoinstall determins if the base iptables rules for cilium to mainly interact with kube-proxy (and masquerading)
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]: "determins" is a typo and it looks like a verb is missing.
Perhaps something more like "IPTablesRulesNoinstall disables installing the base IPTables rules used for masquerading and kube-proxy."

File: pkg/apis/kops/networking.go [hunk 17]
```diff
@@ -195,80 +195,224 @@ const CiliumIpamEni = "eni"
...
+	// Default: false
+	IPTablesRulesNoinstall bool `json:"IPTablesRulesNoinstall"`
+	// AutoDirectNodeRoutes adds automatic L2 routing between nodes
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]:
```suggestion
	// AutoDirectNodeRoutes adds automatic L2 routing between nodes.
```
Perhaps spell out "layer 2"

File: pkg/apis/kops/networking.go [hunk 18]
```diff
@@ -195,80 +195,224 @@ const CiliumIpamEni = "eni"
...
+	// Default: false
+	AutoDirectNodeRoutes bool `json:"autoDirectNodeRoutes"`
+	// EnableNodePort replaces kube-proxy with Cilium's BPF implementation
+	// Requires masquerading to be disabled
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]:
```suggestion
	// EnableNodePort replaces kube-proxy with Cilium's BPF implementation.
	// Requires masquerading to be disabled.
```

File: pkg/apis/kops/networking.go [hunk 19]
```diff
@@ -195,80 +195,224 @@ const CiliumIpamEni = "eni"
...
+	// DisableIpv4 is deprecated: Use EnableIpv4 instead.
+	// Setting this flag has no effect.
+	DisableIpv4 bool `json:"disableIpv4,omitempty"`
+	// DisableK8sServices has not been implemented in the new templates and may be removed in the future.
+	// Setting this has currently no effect.
+	DisableK8sServices bool `json:"disableK8sServices,omitempty"`
+	// EnablePolicy determines the policy enforcement mode.
+	// "default": Follows Kubernetes policy enforcement.
+	// "always": Cilium restricts all traffic if no policy is in place.
+	// "never": Cilium allows all traffic regardless of policies in place.
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]: "" appears to be a supported value.

File: pkg/apis/kops/networking.go (hunk 18 above)
Comment (johngmyers) [reply pkg/apis/kops/networking.go]: Also requires `spec.kubeProxy.enabled` to be false.

File: pkg/apis/kops/networking.go (hunk 14 above)
Comment (olemarkus) [reply pkg/apis/kops/networking.go]: I am wondering if I should remove this part. Cilium 1.7.0 ignores this setting anyway, and the argument will be removed in 1.8.0

File: pkg/apis/kops/networking.go (hunk 16 above)
Comment (olemarkus) [reply pkg/apis/kops/networking.go]: Thanks

File: pkg/apis/kops/networking.go (hunk 13 above)
Comment (olemarkus) [reply pkg/apis/kops/networking.go]: Fixed

File: pkg/apis/kops/networking.go (hunk 19 above)
Comment (olemarkus) [reply pkg/apis/kops/networking.go]: Fixed

File: pkg/apis/kops/networking.go (hunk 18 above)
Comment (olemarkus) [reply pkg/apis/kops/networking.go]: Thanks for catching this. `spec.kubeProxy.enabled` is actually the right one, not  `masquerade`. The latter must be disabled when ipam=eni

File: pkg/apis/kops/networking.go [hunk 20]
```diff
@@ -195,80 +195,226 @@ const CiliumIpamEni = "eni"
...
-	ToFqdnsEnablePoller    bool   `json:"toFqdnsEnablePoller"`
+	// ClusterName is the name of the cluster. It is only relevant when building a mesh of clusters.
+	ClusterName string `json:"clusterName"`
+	// ToFqdnsEnablePoller determines the implementation of FQDN policies.
+	// if this is set to false, the more powerful DNS proxy-based implementation is used.
+	// Enable this option if you want to use FQDN policies but do not want to use
+	// the DNS proxy. To ease upgrade, users may opt to set this option to "true".
+	// Default: false
+	ToFqdnsEnablePoller bool `json:"toFqdnsEnablePoller"`
+	// ContainerRuntimeLabels determines the container runtime(s) used by Cilium
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]:
```suggestion
	// ContainerRuntimeLabels enables fetching of container-runtime labels from the specified container runtime and associating them with endpoints.
```

File: pkg/apis/kops/networking.go [hunk 21]
```diff
@@ -195,80 +195,226 @@ const CiliumIpamEni = "eni"
...
+	// As of Cilium 1.7.0, Cilium no longer fetches information from the
+	// container runtime and this field is ignored.
+	// Default: none
 	ContainerRuntimeLabels string `json:"containerRuntimeLabels,omitempty"`
-	IPTablesRulesNoinstall bool   `json:"IPTablesRulesNoinstall"`
-	AutoDirectNodeRoutes   bool   `json:"autoDirectNodeRoutes"`
-	EnableNodePort         bool   `json:"enableNodePort"`
-	Ipam                   string `json:"ipam,omitempty"`
+	// Ipam determines the IP address allocation mode to use.
+	// "eni" will use AWS native networking for pods. Eni requires masquerade to be set to false.
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]: It looks like the possible values are "crd", "eni", and "azure". These should be documented.

File: pkg/apis/kops/networking.go [hunk 22]
```diff
@@ -195,80 +195,226 @@ const CiliumIpamEni = "eni"
...
 	SidecarIstioProxyImage string `json:"sidecarIstioProxyImage"`
-	ClusterName            string `json:"clusterName"`
-	ToFqdnsEnablePoller    bool   `json:"toFqdnsEnablePoller"`
+	// ClusterName is the name of the cluster. It is only relevant when building a mesh of clusters.
+	ClusterName string `json:"clusterName"`
+	// ToFqdnsEnablePoller determines the implementation of FQDN policies.
+	// if this is set to false, the more powerful DNS proxy-based implementation is used.
+	// Enable this option if you want to use FQDN policies but do not want to use
+	// the DNS proxy. To ease upgrade, users may opt to set this option to "true".
+	// Default: false
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]:
```suggestion
	// ToFqdnsEnablePoller replaces the DNS proxy-based implementation of FQDN policies
	// with the less powerful legacy implementation.
	// Default: false
```

File: pkg/apis/kops/networking.go [hunk 23]
```diff
@@ -195,80 +195,226 @@ const CiliumIpamEni = "eni"
...
+	// Supported values are: "none", "containerd", "crio", "docker", "auto"
+	// As of Cilium 1.7.0, Cilium no longer fetches information from the
+	// container runtime and this field is ignored.
+	// Default: none
 	ContainerRuntimeLabels string `json:"containerRuntimeLabels,omitempty"`
-	IPTablesRulesNoinstall bool   `json:"IPTablesRulesNoinstall"`
-	AutoDirectNodeRoutes   bool   `json:"autoDirectNodeRoutes"`
-	EnableNodePort         bool   `json:"enableNodePort"`
-	Ipam                   string `json:"ipam,omitempty"`
+	// Ipam determines the IP address allocation mode to use.
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]:
```suggestion
	// Ipam specifies the IP address allocation mode to use.
```
(I don't think the word "determines" fits well in this context.)

File: pkg/apis/kops/networking.go [hunk 24]
```diff
@@ -195,80 +195,226 @@ const CiliumIpamEni = "eni"
...
+	// DisableConntrack has not been implemented in the new templates and may be removed in the future.
+	// Setting this has currently no effect.
+	DisableConntrack bool `json:"disableConntrack,omitempty"`
+	// DisableIpv4 is deprecated: Use EnableIpv4 instead.
+	// Setting this flag has no effect.
+	DisableIpv4 bool `json:"disableIpv4,omitempty"`
+	// DisableK8sServices has not been implemented in the new templates and may be removed in the future.
+	// Setting this has currently no effect.
+	DisableK8sServices bool `json:"disableK8sServices,omitempty"`
+	// EnablePolicy determines the policy enforcement mode.
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]:
```suggestion
	// EnablePolicy specifies the policy enforcement mode.
```

File: pkg/apis/kops/networking.go (hunk 21 above)
Comment (olemarkus) [reply pkg/apis/kops/networking.go]: "crd" should work out of the box. Kops wouldn't support azure though.

File: pkg/apis/kops/networking.go [hunk 25]
```diff
@@ -386,6 +386,8 @@ type CiliumNetworkingSpec struct {
 	ContainerRuntimeLabels string `json:"containerRuntimeLabels,omitempty"`
 	// Ipam specifies the IP address allocation mode to use.
 	// "eni" will use AWS native networking for pods. Eni requires masquerade to be set to false.
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go]:
```suggestion
        // Possible values are "crd" and "eni".
	// "eni" will use AWS native networking for pods. Eni requires masquerade to be set to false.
```

**Pull request merged**